Gesture data screen

Helium314 opened the pull request:
@the-eclectic-dyslexic intends to work on a native gesture typing library (#668), ideally usable as a drop-in replacement for the AOSP / HeliBoard native library (could be used similar to the Google library, but no need since it will be open source with a compatible license).
For understanding current handing of gesture typing in the closed source Google library, @the-eclectic-dyslexic asked me to add a screen where users can simply gesture type a word, and the target word plus all available data will be stored.
Data can then be sent via email.

Getting and storing the gesture typing data already works in the current state of the PR, but what data exactly should be stored is not finalized.
The UI... works but should be better.

When sending data via mail, the data file is currently not attached for some reason, possibly an issue with the `provider`.

@the-eclectic-dyslexic if you want to test it, do you compile it yourself, or should I upload the current state as APK?

Helium314 committed:
add screen for gathering gesture data (WIP)

Helium314 committed:
send data via mail (not working yet, nothing attached)

the-eclectic-dyslexic commented:
I will build it and take a look on Monday! Thanks for your work on this @Helium314. I will build it and see what I can think of with regards to the UX of gathering gesture data. I may makes some changes and see what you think before we add it to the apk. Just want to make sure things work as seamlessly as they can before we let people play with it, so that we don't turn anyone off of helping generate data.

Helium314 commented:
> I may makes some changes and see what you think

Sure, we should both agree on the UX and the data (what is included and how it's structured) before merging.

Helium314 committed:
load most of the word list in background

results in first word(s) coming from a small subset of the dictionary, but much better UX

Helium314 committed:
Merge branch 'main' into gesture-data

the-eclectic-dyslexic commented:
Okay! I took a look at it. There are a few things to change for sure, but it should be doable. 

I think the reason the reason the file isn't attaching has something to do with what the provider thinks the file is. I was able to attach it to one email I have an app for on my phone, but not another. In the one where I was able to attach the file, it automatically gave it a .bin extension... even though it was definitely a zip file. I was able to open it after downloading the email. The data was all there! So, it could be that your email client was rejecting files with on extensions or something silly? I will see if I can figure this out.

We should also include the email address by default in the sender field that we want to use. Though, we may want to obfuscate it in the code of heliboard, so it doesn't get crawled and then the inbox gets flooded.

I think we could make a few little changes. Instead of generating a uid for the user, generate one for the gesture. We can talk about this more later! 

I think also maybe we should add the heliboard version string to each gesture, in case they generate data across multiple versions of heliboard. This will come in handy for separating our gesture data from the google library gesture data if we want to keep allowing people to submit data after we distribute our own gesture typing.

I think we should create a cutoff that forces a reentering of a gesture. If a word doesn't show up at all in the suggestions for a gesture, we can assume it isn't correct. I think what we could do is add some kind of indication to the user that we think the input they have given is fine. 

I still need to think a little about the text box. I think it would possibly be better to hide what their gesture resolves to, and only ask them to do well enough that the gesture could reasonably be what we are looking for. I don't want them to be trying to get the exact right gesture, because then all we are doing is optimising for how people work around the failures in gesture recognition, instead of optimising for how they expect it to work intuitively. Basically, we would likely only be able to do as well as the google library with that data... not better. We want them to just kind of wing it like they are trying to write quickly, and see where it is getting the wrong answer, but probably shouldn't have.

I hope all of that made sense. I am going to play with this a bit, but I think this is most of the way there!

Helium314 committed:
add version code

Helium314 committed:
add gesture library hash to data

Helium314 commented:
> So, it could be that your email client was rejecting files with no extensions or something silly? I will see if I can figure this out.

Good point, maybe I just need to change the name in the content URI of the provider.

> We should also include the email address by default in the sender field of the email. Though, we may want to obfuscate it in the code of heliboard, so it doesn't get crawled and subsequently spammed into oblivion.

Definitiely there should be a prefilled address. Maybe having a dedicated address and automatic processing is doable, then spam should be less of an issue.

> heliboard version string

done

> if we are, we may also want to include the gesture library in the version string somehow.

The gesture library sha256 should already be in shared preferences, we could use this to identify the known versions (currently only 1 for each of arm / arm64 / x86 / x86_64).
I added this now, but could also be done differently.

> I think what we could do is add some kind of indication to the user that we think the input they have given is fine. This will take a little bit of changing the UI, but I think I can figure this out.

Would be good to have some ideas, because currently I don't (but it's still early morning...).

> I don't want them to be trying to get the exact right gesture, because then all we are doing is optimising for how people work around the failures in gesture recognition

That would require force-disabling the floating preview setting while in this screen. Technically no problem, but might be unexpected for users who are used to it. Though there is no preview for the final word anyway.

Helium314 commented:
> Good point, maybe I just need to change the name in the content URI of the provider.

No change after appending `.zip` to the URI (using K9 on Android 9)

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
split & move file(s), more share intents + tests

Helium314 committed:
placeholder text and comments

Helium314 committed:
add mail info for file share intent

Helium314 committed:
start the review screen

Helium314 committed:
some update

Helium314 committed:
make review screen work, extend

Helium314 committed:
make selection work

Helium314 committed:
more space for the actual data

Helium314 committed:
give URI permission in intent instead of manifest

Helium314 committed:
remove user-id, more ideas

Helium314 committed:
add logging to content provider

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
fix build

Helium314 committed:
only use the ACTION_SEND intent, but limit it to mail apps

Helium314 committed:
make export to mail more compatible

by @the-eclectic-dyslexic

Helium314 committed:
todo

Helium314 committed:
add missing file

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
add hash to dictionaries, comments in wordData

Helium314 committed:
add db and use it

Helium314 committed:
add missing file

Helium314 committed:
make it work somewhat

Helium314 committed:
tune which data is stored

Helium314 committed:
add ignore list dialog

Helium314 committed:
make sort and peristing ignore list work

Helium314 committed:
improve review screen for landscape mode

Helium314 committed:
continue work on review screen

Helium314 committed:
share / delete what is selected

Helium314 committed:
improve ignorelist handling

Helium314 committed:
improve display

Helium314 committed:
improve sorting

Helium314 committed:
better text for share / delete buttons

Helium314 committed:
adjust filter

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
rename and comments

Helium314 committed:
make review filter a bit nicer

Helium314 committed:
hide the text input field

Helium314 committed:
autosave words

Helium314 committed:
add active / passive column

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
update checksum stuff

Helium314 committed:
change text if other keyboards are used

Helium314 committed:
go from entering gesture data to review screen instead of showing share options

Helium314 committed:
adjust ignore list case treatment

Helium314 committed:
move stuff around in GestureDataScreen

Helium314 committed:
comments

Helium314 committed:
disable floating preview for active gathering

Helium314 committed:
update filtering and when to save gesture data

Helium314 committed:
move data around to more suitable places

Helium314 committed:
tweak data sharing

Helium314 committed:
revert change to donottranslate.xml

Helium314 committed:
save gesture data in coroutine

takes between 20 and 100 ms, which is simply too much otherwise

Helium314 committed:
comments

Helium314 committed:
write json lines instead of a long single lines

Helium314 committed:
don't show irrevelant dictionaries

Helium314 committed:
show word count

Helium314 committed:
minor improvements to data screen

Helium314 committed:
more info about words

Helium314 committed:
add license headers

Helium314 committed:
slightly improve review screen

Helium314 committed:
move passive gathering exclusions into gesturedatascreen

Helium314 committed:
add app ignore list

Helium314 committed:
add filter for app list

Helium314 committed:
improve review screen bottum buttons

Helium314 committed:
comment

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
comments

Helium314 committed:
update documentation and disable passive data gathering

Helium314 committed:
reduced UI in active gathering mode

Helium314 committed:
comment

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
make active gathering work in wide screen mode

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
us isWideScreen

Helium314 committed:
update hashes

Helium314 committed:
store full key size

Helium314 committed:
remove useless json encoding

Helium314 committed:
mark data as exported

Helium314 committed:
show delete only after export

Helium314 committed:
string resources

Helium314 committed:
some more string resources, move sort to icons

Helium314 committed:
add missing icons

Helium314 committed:
scroll to top on sort

Helium314 committed:
delete: return count, optionally delete only exported data

Helium314 committed:
store deleted exported word count

Helium314 committed:
simple bottom bar for active mode

Helium314 committed:
add weight to the active gathering words (use probability / frequency)

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
ask user to upload and clean db if there are more than 10k entries

Helium314 committed:
have expiration date

Helium314 committed:
comment code, move some of it

Helium314 committed:
fix

Helium314 committed:
fix wordlist creation

Helium314 committed:
some strings and mark todos

Helium314 committed:
fixes and better json file formatting, by @the-eclectic-dyslexic

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
comments / reminders for removing data gathering when project is done

Helium314 committed:
more comments / reminders

Helium314 committed:
string resources

Helium314 committed:
don't store gesture lib hash

Helium314 committed:
hide data gathering 2 weeks after end

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
update gesture screen and resource strings

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
more text updates

Helium314 committed:
back to end active gathering

Helium314 committed:
more text and minor adjustments

Helium314 committed:
update readme

Helium314 committed:
rename gesture data json file

Helium314 committed:
Merge branch 'main' into gesture-data

Helium314 committed:
update string

Helium314 committed:
fix issues when quickly switching dictionary

Helium314 committed:
improve dictionary selection

Helium314 committed:
rename property

Helium314 committed:
adjust knownLibrary to boolean?

Helium314 committed:
add notification in case many words are awaiting export

Helium314 committed:
limit exports to 10k words at once

Helium314 committed:
UI adjustments, (dialog arrangements, replace text button)

Helium314 committed:
show message when close to the end of that data gathering phase

Helium314 committed:
move informAboutTooManyPassiveModeWords

Helium314 committed:
add share icon

Helium314 committed:
store popup keys too, and store text instead of code points

Helium314 committed:
disable share & delete buttons after deleting

Helium314 committed:
update texts and set mail address

Helium314 committed:
remove "export" from gesture export strings, use submit / share

Helium314 committed:
clarify that new strings section should be separate

Helium314 committed:
tune active gathering probabilities to favor more common words

Helium314 committed:
update dict hashes

Helium314 committed:
add info to the dict hash file

Helium314 committed:
properly upgrade the DB

Helium314 committed:
fix db creation

Helium314 committed:
remove unfinished parts of passive data gathering (to be added separately)
